### ISSUE TITLE
I2I: Support CMP Integration with <amp-consent>

### ISSUE THREAD

**Issue by zhouyx** [ampproject/amphtml#17742]
To #15653

## Overview ##
The following sections will try to answer these questions:

- What to include in the prompt dialog and how to allow CMP to customize it?
- What is the blocking behavior with CMP?
- What the data flow looks like between different parties?
- What the CMP needs to do to integrate with AMP?
- What does a publisher need to do to include a CMP solution on their AMP page?

## Background ##

[`<amp-consent>`](https://www.ampproject.org/docs/reference/components/amp-consent) provides a solution for publishers to collect user consent and block or unblock any AMP component in an AMP page. Currently, this solution only allows the publisher to collect a boolean consent for all vendors in the consent UI.

To support per purpose and per vendor based consents and allow for third party integration from Consent Management Platforms (CMP) with publisher AMP pages, AMP will enhance `<amp-consent>`. It will allow for richer interaction between the publisher page, vendors, CMP and ad extensions on AMP pages that need to know about the consent status.   AMP must make such an API for CMPs because AMP doesn’t allow custom JavaScript to run on the page and therefore a CMP must integrate with the AMP framework to integrate with a publishers page.

## Data Flow Desing ##
![image](https://user-images.githubusercontent.com/5419904/44685357-23ac0c80-aa00-11e8-8f3a-cdf18858bdfe.png)

The following section explains each step in terms of data structure and API.

#### Read data from localStorage (Step 1) ####
Data includes the stored consent string and CMP stored information.

#### Pass information to CMP endpoint (Step 2) ####
The following data with be passed through a POST request
```
   {
     'consentInstanceId': string,
     'consentStr': string,
     'pubConfig': object,
     'cmpStoredData': string,
   }
```

#### CMP endpoint response (Step 3) ####
Endpoint returns a JSON object
```
   {
     'promptIfUnknown': boolean,
     'sharedData': object,
     'forcePrompt': boolean, //optional
   }
```

#### Initiate the UI iframe (Step 4) ####
Data will be stringified and passed via the iframe name attribute
```
   JSON.stringify({
     'publicConfig': object,
     'cmpStoredData': string,
     'consentStr': string,
   })
```

#### Iframe returns user decision (Step 5) ####
UI iframe will pass the consent information back to `<amp-consent>` via postMessage
```
window.parent.postMessage({
 type: 'consent-response',
 cmpStoredData: '',
 consentStr: '',
 action: 'accept/reject/dismiss'
}, '*');
```
`consentStr` is required to pass the value to vendors.

`action` is optional. When specified, it instructs components that don’t understand consent string to load or not based on the publisher’s configurations.

`cmpStoredData` value will be stored per the CMP request

#### Store consent information (Step 6) ####
Store the `consentStr` and the `cmpStoredData` to localStorage

#### Pass consent information (Step 7) ####
AMP components on the AMP page (vendors in the format `<amp-ad type=${vendorName})>` can access consent information with function call
```
getConsentPolicySharedData(ampdoc, policyId)
```

Where the function returns an object that merge the value of `consentStr` and `sharedData`
```
   {
     'consentStr': string,
     'sharedData': object
   }
```

#### Default Blocking Behaviors ####
When a CMP is present on the page, AMP runtime behavior will default to:
- Block all `<amp-ad>` components on the page from rendering
- Send request to the CMP endpoint to ask whether consent is required
  - If no -> Stop here
- Check if there’s a response stored in the AMP runtime, or the CMP endpoint returned with a response
  - If yes -> Unblock `<amp-ad>` and pass the response to them
  - If no -> Prompt the consent collect dialog in a foreign iframe
- Get the response from the prompted iframe
Store the response as instructed.
- Unblock all `<amp-ad>` components (unless the publisher specifies not to by adding the `data-block-on-consent` attribute) and passesthe response to them.

#### Special Cases ####
Publishers can override the default behavior and customize the blocking behaviors by tagging the components manually.
- If a publisher wants to avoid ad blocking with a CMP,: Use `<amp-ad data-block-on-consent=”_never”>`
- If the publisher wishes to block a component other than `<amp-ad>`: Use the existing `data-block-on-consent` attribute. However,  the value of the returned action (accept/reject/dismiss), depends on the agreement between the publisher and the CMP.

#### External Iframe ####
AMP consent will load and render the CMP’s  prompt UI in a cross domain iframe, where the CMP will have full control. However, AMP runtime will control the life cycle of the iframe, dictating when to load or unload it.
- The iframe will be loaded if a user consent is required but missing, or if a user wants to manage their consent at any point.
- The iframe will be unloaded once a user decision has been made and the prompt UI needs to be hidden.

#### Default placeholder ####
AMP will apply a placeholder before the CMP iframe loads to improve user experience.
The placeholder will not block the user from interacting with the rest of the page.
UX design will be added once ready.

#### Iframe Size and Placement ####
AMP will make the prompt UI interstitial by default. CMP and publishers will NOT be able to customize the size and the placement of the iframe. AMP will create a modal layer and place the iframe with default aspect width and height ratio in the center.

## Consent String ##
Consent string contains the user decision and will be stored and passed to vendors. Vendors will interpret the consent string and respond accordingly.

#### Format ####
Because different CMPs want to store different consent strings for  customized vendors in addition to the IAB consent string, AMP will not enforce format on the consent string, or attempt to interpret the consent string before passing the information to vendors. AMP would require CMP to indicate their string as IAB so that vendors can understand. Please note in MVP, not API will be provided to vendors, and vendors need to interpret consent string themselves. 

#### Storage Limit ####
When the cached AMP page is served from the google.com. All information will be stored under the google.com domain. Because of it, AMP consent has to be very cautious of the storage usage.

In order to be space efficient, AMP restricts the size of total stored value to an upper limit of **300 bytes**. That includes the consent string and the CMP private stored string, if one exists. CMPs should work with their partner vendors to compress their consent string.

In the case that a CMP asks AMP to store information that exceeds the size limit, AMP will:
- Erase the current stored value
- Log an error
- Unblock `<amp-ad>` components and make the consent string information globally accessible to all AMP components the consent string information but w/o storing the value for next visit.

#### Access ####
The consent string is made available to all AMP components on the page and to the CMP.

- To AMP components and vendors

  AMP components can call where the policyId can be interpreted from the data attribute.
  ```
  getConsentPolicySharedData(ampdoc, policyId)
  ```

  3P ad that integrate with AMP can find the string at (similar to supported sharedData)
  ```
  window.context.consentString
  ```

- To CMP
  - CMP will have access to the consent string. Per their request a CMP will also have access to the CMP stored data. Note that passing CMP stored data is not in scope with the initial implementation.
  - The stored value will be divided into two parts.
    - Standard Consent string
    - Private CMP stored string

     Where consent string can be shared to all components, stored string will only be shared with the CMP.

## CMP Integration ##

A CMP named `foo_bar` will integrate with AMP in the following way.

#### Check in configuration ####
Add the configuration as part of the `<amp-consent>` codebase. This will allow AMP to recognize the CMP, perform steps to check if consent is necessary, handle storage and UI.
```
 'foo_bar': {
   ‘storageKey’: ‘cmp_foo_bar’,
   'checkConsentHref': 'https://foo-bar.com/amp/endpoint',
   'promptUISrc': 'https://foo-bar.com/cmp.html',
  }
```
`storageKey` instructs AMP runtime to store the consent information under this unique key name. Note: AMP will append prefix to the key. If not specified, AMP will assume CMP does not require consent string to be stored client side.

Unlike the current DIY approach with `<amp-consent>`, CMP will be able to specify the storage key value instead of publishers. This is easier for CMP to bump up versions because they have control of the storageKey.

`checkConsentHref` is a field AMP consent already supports. It allows the AMP consent component to ask the CMP if consent is required for a user. This field is required, as it is the key to instruct AMP runtime to ask for consent and block other components on the page. For more details, please refer to the section [here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/amp-consent.md#checkconsenthref)

Note:the existing `promptIfUnknowForGeoGroup` for CMP integration won’t be provided as AMP expects CMP to host their own endpoint instead of depending on publishers configuration on the `<amp-geo>` component.

`src` is the source for the foreign iframe. It is also a required field. Once AMP consent decides to prompt, it will create a foreign iframe with the src given.

#### Host endpoint ####
AMP consent will rely on the host endpoint to determine if the prompt is required with the `checkConsentHref` request.

#### Consent Dialog ####
The CMP will need to modify their prompt logic to function in a cross domain iframe. And communicate with the parent host AMP page through the proposed communication APIs.

## Publisher’s Configuration ##
To work with a CMP, publisher’s configuration will need to stay minimal.
```
   <amp-consent id="ABC" type="foo-bar" layout="nodisplay">
     <script type="application/json">{
       “cmp”: {
         “token”: xxx
       }
     }</script>
   </amp-consent>
```

#### Type Attribute ####
A type attribute with a valid CMP name value instructs the `<amp-consent>` component to load the CMP config.

#### CMP Object ####
All configuration within the `cmp` object will be passed to the CMP, both with `checkConsentHref` and iframe (added to the name attribute).

#### PostPromptUI ####
CMP’s support to postPromptUI has not been discussed yet. Right now publishers can still put their own postPromptUI element on the page and reference the element.
```
   <amp-consent id="ABC" type="foo-bar" layout="nodisplay">
     <script type="application/json">{
       "postPromptUI": "postPromptUI"
     }</script>
     <div id="postPromptUI">
       Post Prompt UI
       <button on="tap:ABC.prompt" role="button">Manage</button>
     </div>
   </amp-consent>
```

## Other topics ##
Below is a list of topics we have talked about but not covered in the current design. We may look into supporting it in the future.
Allow publishers to still use the `<amp-geo>` component to reduce one roundtrip of latency
Provide support for CMP to store consent server side but not client side

**Comment by jasti** [comment]
Hi @jeffnordlund-ghost @jawadst @janwinkler, @waschnick and @consentmanager. 
Since you've previously expressed interested in amp-consent's ability to integrate with CMPs, we'd love to have you participate in our design review this Wed @ 1PM PT where we'll go over this component and it's design. Details in #17357.
Please feel free to leave feedback on this issue in the mean time.

**Comment by jawadst** [comment]
@jasti Thanks for pinging me, I'll definitely join!

**Comment by zhouyx** [comment]
Few thing I recall from the design review

1. The concern about blocking analytics vendors
    - There's no good way to support that with the current design
    - Would require analytics offline mode to store events before user giving consent
    - To follow up with some analytics vendors and see if they wish to be blocked. 
2. The concern about vendor list changes
     - To introduce dirty bit to force ask user consent the next time they visit the page
3. Decided to support the integration with `<amp-geo>` with CMP integration as it provides faster experience for users that doesn't require consent. 
4. storage limit concern
     - CMP use the client side storage as their primary source of storage. 
     - We'll start with 300 bytes limit and work on the LRU/LRR to expand the storage limit.

**Comment by lannka** [comment]
Hi @jeffnordlund-ghost @jawadst @janwinkler, @Waschnick and @consentmanager

we are looking for a CMP to integrate with the amp-consent API. Please let me know if you are interested and we can schedule a VC to go over integration details.

**Comment by consentmanager** [comment]
yes please :-)

**Comment by oliverfernandez** [comment]
Hi @lannka 

We also have an IAB certified CMP, deployed in several sites, what we'd like to integrate with AMP. I'd appreciate a lot if you could keep me in the loop regarding this

Thanks!

**Comment by jawadst** [comment]
@lannka Happy to help! We already have a first setup running with the existing amp-consent at Didomi so happy to see how the new version would perform with that.

**Comment by jeffnordlund-ghost** [comment]
Sure, we would like to explore a better integration with the amp-consent api.

**Comment by lannka** [comment]
Thanks all.  @consentmanager @oliverfernandez @jawadst @jeffnordlund-ghost 
If you have a launched CMP product, could you share a link?

Also what's your timezone so we can schedule a VC.

**Comment by oliverfernandez** [comment]
You can find and example of our CMP working here (mobile only): https://nospensees.fr/

My timezone is CET as well

**Comment by zhouyx** [comment]
@consentmanager @oliverfernandez @jawadst @jeffnordlund-ghost Thank you all for the interest. We scheduled a meeting at Thursday. 
Time: [2018-09-13 15:30 UTC](https://www.timeanddate.com/worldclock/meeting.html?year=2018&month=09&day=12&iv=0) ([add to Google Calendar](http://www.google.com/calendar/event?action=TEMPLATE&text=amp-consent%20CMP%20integration&dates=20180913T153000Z/20180913T163000Z&details=https%3A%2F%2Fhangouts.google.com%2Fhangouts%2F_%2Fgoogle.com%2Fkky-qrmj-svk))
Location: [Video conference via Google Hangouts](https://hangouts.google.com/hangouts/_/google.com/kky-qrmj-svk)

To #17983 for meeting details. Please let us know. Thanks.

**Comment by oliverfernandez** [comment]
@zhouyx works for me

**Comment by lannka** [comment]
@consentmanager @jawadst @jeffnordlund-ghost in case you didn't see.
are you able to join us tmr?

**Comment by consentmanager** [comment]
Yes, will be there

**Comment by jawadst** [comment]
@lannka 
That works for me as well

**Comment by jeffnordlund-ghost** [comment]
Sorry, I’ve been out of the office for a bit and just catching up.

I might be able to join (if it hasn’t happened yet).  My timezone is MST (in the US).  I thought I saw a mail with the date/time earlier but I’m not seeing it now.  When are you looking to schedule this?

Thanks!
- Jeff

**Comment by zhouyx** [comment]
@consentmanager @oliverfernandez @jawadst @jeffnordlund-ghost 
Thanks for joining the meeting last week. It was really helpful! We would like to schedule a 30 mins follow up meeting this week to talk about technical details and timeline. Thanks. 

Time: [2018-09-20 15:30 UTC](https://www.timeanddate.com/worldclock/meeting.html?year=2018&month=09&day=20&iv=0) ([add to Google Calendar](http://www.google.com/calendar/event?action=TEMPLATE&text=amp-consent%20CMP%20integration&dates=20180920T153000Z/20180920T160000Z&details=https%3A%2F%2Fhangouts.google.com%2Fhangouts%2F_%2Fgoogle.com%2Ftgg-rpqh-ttp))
Location: [Video conference via Google Hangouts](https://hangouts.google.com/hangouts/_/google.com/tgg-rpqh-ttp)

Please let us know if you plan to attend or suggest a different time. Thanks.

**Comment by jawadst** [comment]
@zhouyx There is an IAB Techlab call at that time that a lot of us are likely to attend. Can we do Friday instead?

**Comment by zhouyx** [comment]
Thank you @jawadst for the heads up. Would the same time on Friday, 2018-09-21 15:30 UTC work for everyone? If so I'll go ahead and change the invitation. 

Please let us know. Thanks

**Comment by jawadst** [comment]
Works for me

**Comment by consentmanager** [comment]
works

**Comment by consentmanager** [comment]
re: "CMP and publishers will NOT be able to customize the size and the placement of the iframe. AMP will create a modal layer and place the iframe with default aspect width and height ratio in the center."
--> why the downgrade? currently we are able to resize the iframe, which is necessary when many information is shown ...

**Comment by zhouyx** [comment]
@oliverfernandez  @jeffnordlund-ghost in case you didn't see, we scheduled a meeting Friday to talk discuss the integration work. Thanks 
 
Time: [2018-09-21 15:30 UTC](https://www.timeanddate.com/worldclock/meeting.html?year=2018&month=09&day=21&iv=0) ([add to Google Calendar](http://www.google.com/calendar/event?action=TEMPLATE&text=amp-consent%20CMP%20integration&dates=20180921T153000Z/20180921T160000Z&details=https%3A%2F%2Fhangouts.google.com%2Fhangouts%2F_%2Fgoogle.com%2Ftgg-rpqh-ttp))
Location: [Video conference via Google Hangouts](https://hangouts.google.com/hangouts/_/google.com/tgg-rpqh-ttp)

**Comment by jawadst** [comment]
@zhouyx Apologies as I missed the call. Let me know if you need anything from us.

**Comment by jeffnordlund-ghost** [comment]
Sorry, looks like I missed the call.  Please let me know if I can provide any information or answer any questions from how we are using the IAB.

**Comment by zhouyx** [comment]
Thank you all for attending the meetings. 

Here's the notes from the meetings we have

9/13 Understand the requirement
1. Consent string standard
- Supposedly all ad vendors will eventually support the IAB standard. Analytics vendors however don’t have incentive to do so. 
- Vendor configuration is done on CMP server side
- AMP support the IAB use case. And also the global boolean consent decision (accept/reject) so publishers can work with CMP to block vendors that doesn't support IAB. 

2. UI standard:
- All CMP provides blocking and non blocking version of the UI. (Blocking UI is an interstitial that takes the full screen. Non blocking UI stick to the bottom of the page)
- Publishers have the final word on UI design. And the configuration is from client side. 
- Don’t have good data on which one is preferred by publishers today.
- AMP will by default support UI, that is non blocking before iframe load.

**Comment by zhouyx** [comment]
9/21 Discuss technical details 

1. UI
- AMP will provide iframe ready API that show the iframe and hide placeholder. 
- Initial size is always 30vh, and can’t be changed **Is this OK with everyone??**
(Feedback from the meeting: 30 vh may not be enough because clients can defined the text. Accept button is usually below the text. Also text on the button can be very long with certain languages. But 30vh is fine to start with. Also there's not much extra work to support dynamic sizing from CMP side). 
- AMP offers the requestFullScreen API, that expand the iframe to fullsize for "learn more session". 
- When `postPromptUI` is clicked to trigger the UI again. AMP will toggle the iframe to the initial 30vh size again because not all publishers allows custom settings and it doesn't make sense to jump to fullscreen. 

2. Remote Config

Data from AMP: **Is this sufficient to make the decision?? **
- consentInstanceId. Which is defined by CMP and checked in to AMP code base
- Cmp configuration that is defined by publisher. Note AMP won’t enforce format to the configuration. 

Response from CMP
- If consent is required or not.
- To force clean the stored consent the NEXT time the user visit the page
- sharedData that can be accessed by vendors, but won’t be stored

3. PromptUI Iframe

- To render the iframe all data will be written to the name attribute

Data from AMP **Is this sufficient to render the UI??**
- Stored consent string
- Cmp configuration that is defined by publisher. Same to Remote Config 

APIs
- UI APIs
- Consent-response API (pass back a consent string that will be stored and an optional global boolean decision). 

Update on timeline from AMP
- We're in the middle of the UI support
- Configuration support is already there
- Need to add API support afterwards

We're expecting to have CMP develop in parallel with us and have an end to end integration flow soon. Please let us if you have any question or concerns on the design. And would love to start email thread or github thread to track the development process. Thank you!

**Comment by zhouyx** [comment]
Hi all, Happy New Year. 

We're close to the code complete of the feature on AMP side. Would like to look for early adopters who would integrate with us. Please let us know if you're interested. 

Implementation Progress from AMP

- [x] Stored consent string
- [x] Cmp configuration integration 
- [x] Consent-response API to pass consent string back to AMP runtime
- [x] Render iframe provided by cmp to collect consent. 
- [x] Pass stored consent string and client config to iframe
- [x] Full screen API
- [x] Enable ad vendors to retrieve consent string passed from CMP

One last piece of work we are still working on is to block all `amp-ad` on the page by default. The work can be tracked at #20253. We've decided to introduce a meta tag and require that from publishers in order to block all ads by default. Please let us know if there's any feedback. Thank you. 

to @consentmanager @oliverfernandez @jawadst @jeffnordlund-ghost

**Comment by jawadst** [comment]
@zhouyx Happy to start the integration with the Didomi CMP whenever you are ready. Do you plan to release a test version before the final production release?

**Comment by zhouyx** [comment]
@jawadst great! We'll keep you updated once the last piece of work is done. 

We are protect the feature with an `amp-consent-v2` experimental flag. And plan to flip that flag once we believe it's ready.

**Comment by zhouyx** [comment]
@consentmanager @oliverfernandez @jawadst @jeffnordlund-ghost  The AMP side work has completed. We expect all feature to be live in canary next Tuesday.

I've created a draft doc for you to start with. Please let us know if there's any questions. I'm happy to sync with you offline as well.

**Comment by zhouyx** [comment]
All features should be live in with developer opt-in now. 

I've created an [example page](https://zhouyx-flying-carpet.firebaseapp.com/manual/amp-consent-cmp.html) to demonstrate. It's a simplified cmp service, with

- example remote endpoint logic
![image](https://user-images.githubusercontent.com/5419904/51721636-82090580-2007-11e9-87dc-e6d77920571e.png)

- consent prompt ui iframe
https://raw.githubusercontent.com/ampproject/amphtml/master/test/manual/diy-consent.html

- and configuration in AMP consent
https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/0.1/cmps.js#L30-L36

Please let me know what I can help with the integration. Thank you.

@consentmanager @oliverfernandez @jawadst @jeffnordlund-ghost

**Comment by jawadst** [comment]
@zhouyx Thanks a lot for all your hard work, this looks great!
I'll try it out next week and send any feedback I have.

**Comment by christian-felix** [comment]
Great, thank you @zhouyx. We are also very impatient to use this new solution and adapt it soon.

**Comment by lannka** [comment]
hi @jawadst did you get chance to look at the integration yet?
we're looking for an announcement in the coming [AMP Conf](https://www.ampproject.org/amp-conf/) in April. It will be good if we could show case a couple of CMP integrations.

Let us know if you have any questions.

**Comment by christian-felix** [comment]
Hi @lannka, we are currently evaluating the idea and hopefuly  absorbing it soom.  In case of successfully integration I will provide you with the appropriate link for the coming AMP Conf.

**Comment by jasti** [comment]
Hi @jawadst @consentmanager @oliverfernandez  @jeffnordlund-ghost we are looking to exist the beta relatively soon and would love your feedback before we do so. 
Do you have an ETA for when you might be ready with your AMP CMP integrations?

**Comment by jawadst** [comment]
@jasti @lannka @zhouyx 
We are finishing up our tests and should be able to provide feedback some time next week

**Comment by jawadst** [comment]
@zhouyx  @jasti 
We have a few questions for you following our initial tests:

1) It seems that the function `{amp-consent-id}.prompt()` can be used to re-open the prompt UI, is that correct? Is that documented somewhere yet? (I am assuming it is new with your v2 update)

2) For the `info` field that gets returned by a CMP in the `consent-response` postMessage, we are not completely sure of what you expect in there. I understand it is a general field for a CMP to store data and:
 - Is it expected to be an IAB consent string or is it a free-form field for a CMP to use?
 - Is it the value that vendors get when they ask for the consent string?
 - Is it the value that gets passed back in the `consentString` field in the iframe name? If yes, you might want to rename the fields to have the same name (`consentString` or `iabConsentString` are good).

3) What is your recommended solution for testing the prompt UI iframe and checkConsentHref locally before adding our CMP to https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/0.1/cmps.js? (for local development and testing basically).
We are able to replace your URLs when using your test file with a browser extension but is there a way to locally define a CMP or override the settings for your `_ping_` CMP?

4) Regarding the co-existence of the consent string and the consent state, is our understanding correct that:
- the consent state is used for blocking behaviors as described on https://www.ampproject.org/docs/reference/components/amp-consent#blocking-behaviors
- the consent string will be passed to IAB vendors as needed but is not leveraged by AMP itself for on-page blocking

5) Do you have any documentation on how vendors are expected to query/receive the IAB consent string? Are you in discussion with any vendor on that topic?

6) One open question that I know we discussed on our calls and you decided not to include in this first update to AMP: do you have plans to add more granularity to the consent state and the blocking behaviors on the page?
As it is defined today as a binary state (agree/disagree), it is impossible to really be compliant with GDPR and the latest information given by DPAs like the CNIL at the end of 2018, as they require granular choices (by purpose and by vendor) to be exposed to the user. A global yes/no choice is not enough from that perspective.
While we might be able to achieve that for IAB vendors with your current implementation, that will not be doable for non-IAB vendors like analytics vendors.

Thanks a lot for all your hard work on this subject: we get daily requests from our publishers for AMP support so be assured that this will be used as soon as it is released!
I think we'll have our first AMP CMP released some time in March so let me know if you want to coordinate something for your April conference, I am happy to provide some material as needed.

cc @MaximePlancke (who is also working on our AMP integration at Didomi)

**Comment by zhouyx** [comment]
@jawadst Thank you for the feedback.

1.
> It seems that the function `{amp-consent-id}.prompt()` can be used to re-open the prompt UI, is that correct? Is that documented somewhere yet? (I am assuming it is new with your v2 update)

This is something we support in v1 : ) Document is [here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/amp-consent.md#post-prompt-action). It is designed for users to manage their consent after initial interaction with the prompt.

2.
> * Is it expected to be an IAB consent string or is it a free-form field for a CMP to use?

AMP expect it to be a free-form field. Whether to follow the IAB consent string format or not is up to the CMP. 

>* Is it the value that vendors get when they ask for the consent string?

Exactly


> * Is it the value that gets passed back in the `consentString` field in the iframe name? If yes, you might want to rename the fields to have the same name (`consentString` or `iabConsentString` are good).Is it expected to be an IAB consent string or is it a free-form field for a CMP to use?

You're right! There are other client information that get passed to the iframe in the iframe name. Which includes `clientConfig`, `consentState` and `consentString`. And we pass the stringified  object. Which field do you want to rename?

3. 
> What is your recommended solution for testing the prompt UI iframe and checkConsentHref locally before adding our CMP to https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/0.1/cmps.js? (for local development and testing basically).
   We are able to replace your URLs when using your test file with a browser extension but is there a way to locally define a CMP or override the settings for your `_ping_` CMP?

It is not documented, but you can use inline config to override any CMP provided vendor config. In this case I assume you can do
```
<amp-consent type='_ping_'>
  <script type="application/json">
    {
      'consentInstanceId': yourid,
      'checkConsentHref': ....,
      'promptUISrc': xxxx,
    }
  </script>
</amp-consent>
```
let me know if this helps with your local testing. 

4. 
> Regarding the co-existence of the consent string and the consent state, is our understanding correct that:
>* the consent state is used for blocking behaviors as described on https://www.ampproject.org/docs/reference/components/amp-consent#blocking-behaviors
>* the consent string will be passed to IAB vendors as needed but is not leveraged by AMP itself for on-page blocking

Exactly

5. 
> Do you have any documentation on how vendors are expected to query/receive the IAB consent string? Are you in discussion with any vendor on that topic?

Thank you for asking this. This is a great question. AFAIK no, I'll sync with the team and update with more info tomorrow.

6. 
>One open question that I know we discussed on our calls and you decided not to include in this first update to AMP: do you have plans to add more granularity to the consent state and the blocking behaviors on the page?
    As it is defined today as a binary state (agree/disagree), it is impossible to really be compliant with GDPR and the latest information given by DPAs like the CNIL at the end of 2018, as they require granular choices (by purpose and by vendor) to be exposed to the user. A global yes/no choice is not enough from that perspective.
    While we might be able to achieve that for IAB vendors with your current implementation, that will not be doable for non-IAB vendors like analytics vendors.

This is a great question. And as you've pointed out this is achievable for IAB vendors. To support more granular consent blocking for non-IAB vendors, it requires AMP to come up with a consent string standard and it's not something we are planning to do. However if that becomes a popular request we will definitely consider adding more granularity. 

Thanks again for the questions and feedbacks. Let me know there's more I can help with.

**Comment by janwinkler** [comment]
One question regarding the examples above:

window.parent.postMessage({
 type: 'consent-response',
 cmpStoredData: '',
 consentStr: '',
 action: 'accept/reject/dismiss'
}, '*');

Is this example still correct or does the message need to be a json string instead?

**Comment by zhouyx** [comment]
@janwinkler Good point. The example still applies, except we change the `consentStr` to `info` and the `cmpStoredData` is currently not supported.

**Comment by dmdabbs** [comment]
> We'll start with 300 bytes limit and work on the LRU/LRR to expand the storage limit.
Consent strings are likely to be larger than this within the next year as the IAB TCF version 2 begins deploying.

> Currently we haven't documented on how vendors are expected to query/receive the IAB consent string yet. But it's definitely we will be adding soon.
Perhaps I'm misunderstanding here. (Ad tech) vendors need access to the consent string to send in their ad fetch calls so that they know whether the proper disclosures were made to the user and whether they have consent (if they're relying on consent). Not sure I understand the point of a CMP that generates consent strings to which vendors have no way to access.

**Comment by jawadst** [comment]
@zhouyx 
Thanks for your answers! A few extra comments/questions for you:

1. On the format that is stored by AMP in the `consentString` field (consent string vs CMP-specific string): if what is stored gets passed to vendors then it forces CMPs to store the IAB consent string there and prevents us from storing consent for non-IAB vendors.
I think we'll have to assume that non-IAB vendors are not supported in this first version but please keep them in mind for future evolutions. They are a big part of what we do with our publishers. One option would be to open another field (`cmpConsent`) that would only contain CMP-specific information and would allow us to store extra information as needed.

2. Do you have a way to exit the fullscreen mode?
We sometimes display a banner notice (so the consent UI is not fullscreen initially) with a link to open more information in a pop-in. The pop-in uses the fullscreen mode when it opens which works great. When the user closes the pop-in and if they have not made a consent choice yet, we want to exit the full-screen mode and leave the banner notice open. It does not seem like there is an action to exit the full-screen mode.

3. CSS for iframe
Would it be possible to remove the CSS that has a visual impact (border, for instance) on the prompt UI iframe? I think that there should not be anything else than positioning/height applied.
Right now, it prevents us from styling the UI exactly how we want.

4. Consent re-collection
There are a few cases when consent needs to be recollected and the consent UI needs to be shown again after the user gives consent:
- Consent is more than 13 months old (ie consent expiration)
- A new vendor is added to the list of vendors
And publishers/CMPs usually add extra logic there (a minimum window to never display the consent UI more than once every X days, for instance).
How are those cases managed currently? Ideally, the promtpUI iframe would always be loaded in a hidden state and it would have a special action to let AMP know if consent needs to be collected or not. The iframe/CMP would then be in charge of deciding whether consent needs to be collected or not (which would also allow us to apply extra logic as needed). That would effectively replace the checkConsentHref call with communication with the iframe.
It could also be done with the checkConsentHref call if that call was done on every single page load (and not just when consent is unknown) and that call included the same information as what is passed to the iframe (consentString, consentState, etc.)

---

To summarize, I think the following two points are blockers at this point:

- We need more clarity on how IAB vendors will get passed the IAB consent string and we need some vendors buy-in. I think this might be the right timing to join one of the IAB Techlab meetings for a presentation to get some vendors involved. Happy to help facilitate that if needed.
 - Consent recollection is absolutely required. Publishers add a lot of new vendors regularly and we need a way to define when the consent UI should be shown again.

**Comment by zhouyx** [comment]
@dmdabbs 

>Consent strings are likely to be larger than this within the next year as the IAB TCF version 2 begins deploying.

Thanks for bringing it up. It's definitely something we keep in mind. We are happy to explore ways to loose the restrictions then. 

>Perhaps I'm misunderstanding here. (Ad tech) vendors need access to the consent string to send in their ad fetch calls so that they know whether the proper disclosures were made to the user and whether they have consent (if they're relying on consent). Not sure I understand the point of a CMP that generates consent strings to which vendors have no way to access.

I'm completely with you on this point. The support is already in, the documentation work is deferred until we have the CMP support.

**Comment by zhouyx** [comment]
@jawadst Thank you for the additional feedback.

> 1. On the format that is stored by AMP in the `consentString` field (consent string vs CMP-specific string): if what is stored gets passed to vendors then it forces CMPs to store the IAB consent string there and prevents us from storing consent for non-IAB vendors.
   I think we'll have to assume that non-IAB vendors are not supported in this first version but please keep them in mind for future evolutions. They are a big part of what we do with our publishers. One option would be to open another field (`cmpConsent`) that would only contain CMP-specific information and would allow us to store extra information as needed.

If I understand it correctly, what you want is to store a piece of information that's only for CMP use. This is something we can support as long as the string size under the cap. What confuses me a bit is How will this string work with non-IAB vendors? 

> 2. Do you have a way to exit the fullscreen mode?

Short answer: No, but we can definitely add support

> 3. CSS for iframe

Sounds reasonable to me. @torch2424 who is the owner of amp-consent UI can help answer the question better. 

> 4. Consent re-collection

Thanks for bringing this up. AMP decides to support a one behind model that allows the endpoint to erase the stored consent and force prompt next time. This will be supported with the 'forcePromptOnNext' feature. However it is something the team decides to de-prioritize until CMP integration is added. I'm happy to prioritize if you find it to be a big blocker. 

> We need more clarity on how IAB vendors will get passed the IAB consent string and we need some vendors buy-in. I think this might be the right timing to join one of the IAB Techlab meetings for a presentation to get some vendors involved. Happy to help facilitate that if needed.

I wonder in the non AMP case, how CMP passes the IAB consent string to vendors? I assume it will be very similar in AMP pages. Happy to discuss this further. cc @jasti

**Comment by jawadst** [comment]
@zhouyx 

> 1. On the format that is stored by AMP in the `consentString` field (consent string vs CMP-specific string): if what is stored gets passed to vendors then it forces CMPs to store the IAB consent string there and prevents us from storing consent for non-IAB vendors.
>    I think we'll have to assume that non-IAB vendors are not supported in this first version but please keep them in mind for future evolutions. They are a big part of what we do with our publishers. One option would be to open another field (`cmpConsent`) that would only contain CMP-specific information and would allow us to store extra information as needed.

>  If I understand it correctly, what you want is to store a piece of information that's only for CMP use. This is something we can support as long as the string size under the cap. What confuses me a bit is How will this string work with non-IAB vendors?

What we usually do for non-IAB vendors on the regular web or mobile apps is the following:
- If the vendor has a custom API for consent, we call into that with a custom integration. DFP, for instance, exposes the googletag.pubads().setRequestNonPersonalizedAds function (https://developers.google.com/doubleclick-gpt/reference#googletag.PubAdsService_setRequestNonPersonalizedAds) that can be called to request personalized or non-personalized ads.
 - If the vendor does not have a custom API, we simply condition the loading of the vendor tag to the consent given by the user through a tag manager or a hard `if` in JavaScript

I think that, for AMP, only the second case matters and it could translate into a custom format stored by the CMP like:
`{
  vendorId: true/false,
  vendorId2: true/false,
}`

that could be interpreted by AMP and used in conjunction with a `data-block-on-vendor-consent=""vendorId""` attribute to block any element based on specific consent given to a vendor.
That's just an example of how that would work but the overall idea is that the CMP can report consent status by vendor (instead of a global consent state) and that blocking behaviors can tap into the status by vendor for a more granular control.

> 2. Do you have a way to exit the fullscreen mode?

> Short answer: No, but we can definitely add support

That would be great :)

> 4. Consent re-collection

> Thanks for bringing this up. AMP decides to support a one behind model that allows the endpoint to erase the stored consent and force prompt next time. This will be supported with the 'forcePromptOnNext' feature. However it is something the team decides to de-prioritize until CMP integration is added. I'm happy to prioritize if you find it to be a big blocker.

It would definitely be great to re-prioritize this. Is my understanding correct that the HTTP endpoint defined for `checkConsentHref` is the one that you are refering to when saying "allows the endpoint to erase the stored consent and force prompt next time"? If yes, are you also planning to pass all the consent information (what is being passed to the iframe) when calling that endpoint?

> I wonder in the non AMP case, how CMP passes the IAB consent string to vendors? I assume it will be very similar in AMP pages. Happy to discuss this further.

The IAB spec currently has three different use cases for passing consent to vendors:

- In web pages (web or desktop), the CMP exposes a JavaScript on the page (a `__cmp` function) that vendors are responsible for calling to collect the consent status and act accordingly: https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/CMP%20JS%20API%20v1.1%20Final.md#CMP-JS-API
 - In mobile apps, the CMP writes the consent string into the local storage (NSUserDefault or SharedPreferences) and vendors' SDKs read from there: https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/Mobile%20In-App%20Consent%20APIs%20v1.0%20Final.md
 - For other cases, the consent string can be passed as a query-string parameter with somewhat standardized parameters and macros; https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/URL-based%20Consent%20Passing_%20Framework%20Guidance.md (and the future version of it proposed by Google: https://docs.google.com/document/d/1laqa_jilzDxDoiF-341PLP7QM_NEXqUmbwgQ7U_JS54/edit#heading=h.nunn6u7p2k25)

I am not an AMP expert so I am not sure what would work for you. Maybe the #3. I think it would also be acceptable to have an AMP-specific way as every platform tends to have its own preferred way of passing data around.

**Comment by torch2424** [comment]
@jawadst Hello, sorry for cutting in here, but yes we can definitely do the CSS thing. If you could, can you provide a small example of the iframe contents you are looking to achieve. Perhaps even a screenshot, and I can look into a fix for you 😄

**Comment by zhouyx** [comment]
@torch2424 Let's track the UI issue in a separate thread : )

@jawadst thank you for the detailed explaination
>I think that, for AMP, only the second case matters and it could translate into a custom format stored by the CMP like:
`{ vendorId: true/false, vendorId2: true/false, }`
that could be interpreted by AMP and used in conjunction with a `data-block-on-vendor-consent=""vendorId""` attribute to block any element based on specific consent given to a vendor.
That's just an example of how that would work but the overall idea is that the CMP can report consent status by vendor (instead of a global consent state) and that blocking behaviors can tap into the status by vendor for a more granular control.

Ad vendor specific blocking can be achieved today. This would require the ad vendor to add their support to AMP. We expose the `getConsentState()` to them. However others like analytics vendor is currently not available because analytics vendors only check in the configuration json to AMP, and can't handle consent status client side today. The AMP team has considered the case here. AMP doesn't want to enforce any consent string standard, so if there's future request for blocking individual vendors, we can explore other approaches such as allowing integrated vendors to insert their plugin functions. 

>I am not an AMP expert so I am not sure what would work for you. Maybe the #3. I think it would also be acceptable to have an AMP-specific way as every platform tends to have its own preferred way of passing data around.

Yes. Having the consent string passed as a query-string parameter with somewhat standardized parameters and macros is what AMP supports today. So I assume vendors can handle the consent string from AMP pages without much of extra effort. 

>It would definitely be great to re-prioritize this. Is my understanding correct that the HTTP endpoint defined for `checkConsentHref` is the one that you are refering to when saying "allows the endpoint to erase the stored consent and force prompt next time"? If yes, are you also planning to pass all the consent information (what is being passed to the iframe) when calling that endpoint?

I'll add the task to our spring planning list. Correct it's the same endpoint I'm referring to. I suppose the consent information is what's required to determine to force prompt or not, so yes if it is requested.

**Comment by jawadst** [comment]
@torch2424 
Thanks for chiming in!

Here is an example of what the current style is with a fullscreen pop-in:
<img width="1440" alt="capture d ecran le 2019-03-01 a 19 30 17" src="https://user-images.githubusercontent.com/584597/53660543-17fa0680-3c2c-11e9-8dcf-c0edd5a8a84b.png">

As you can see, there is a weird border at the top of the screen. We want to make sure there is no background, border or or box shadow applied to the iframe that contains the CMP UI (for both fullscreen and non-fullscreen mode0.

**Comment by torch2424** [comment]
@jawadst Awesome, thanks for sending that!

Opened an issue at: #21245

**Comment by MaximePlancke** [comment]
Hi @torch2424 @zhouyx ,

We have worked on our Iframe and POST endpoint. I created a PR to add Didomi to the list of vendors https://github.com/ampproject/amphtml/pull/21693

I wasn't sure which branch or what to reference. Please let me know if I need to do any changes.

Thank you

**Comment by jawadst** [comment]
@torch2424 @zhouyx 
Do you have a sense of when the updates to AMP Consent will be merged into master and the documentation updated? (https://amp.dev/documentation/components/amp-consent)

We are ready to start deploying a few of our publishers and would like to know what timeline we can communicate to them.

I also want to re-emphasize the importance of working out how the IAB consent string gets passed to vendors :)

**Comment by torch2424** [comment]
@jawadst 

Hello! Thank you for your patience, we were out for AMP Conf last week.

If I could ask, which updates are you looking for? Can you pass along some PRs you are waiting to be deployed? That way I can help you from there.

But in terms of my Consent PRs I may have made, those should be in PROD, but there are some pending documentation details I need to add.

Let me know if there is a particular PR. Thanks! 😄 👍

**Comment by jawadst** [comment]
@torch2424 
We weren't clear on whether the support for CMPs in <amp-consent> was deployed in AMP PROD as we have been testing with a custom build. Is it still part of an experiment? ie do our clients need to add the following code to their pages:

```
AMP.toggleExperiment('amp-consent-v2', true);
```

The main features that were discussed earlier in this issue and that we think are high priority:
- Being able to exit full-screen mode from the iframe (ie have a postMessage to tell AMP to exit full-screen mode). In our current setup, once the user enters full-screen mode (clicking on "Learn More" from a notice, for instance), we are unable to exit it.
- Pass the consent string to IAB vendors. Currently, no matter what choices the user makes, IAB vendors do not get the consent string and are unable to respect the user's choices.
 - Pass the CMP configuration to checkConsentHref endpoint

Do you have Github tickets to track those?

**Comment by torch2424** [comment]
@jawadst Thanks for the thorough response. 😄 

I'll let @jasti chime in here, as they are the PM for `<amp-consent>`. 

But to answer some of your questions:

> Is it still part of an experiment? ie do our clients need to add the following code to their pages

Yes, there is still an experiment for our CMP implementation of `amp-consent`. I can sync with @zhouyx and @jasti on that 👍 

> The main features that were discussed earlier in this issue and that we think are high priority:=

Awesome thank you for this! I opened the following issues:

> Being able to exit full-screen mode from the iframe (ie have a postMessage to tell AMP to exit full-screen mode). In our current setup, once the user enters full-screen mode (clicking on "Learn More" from a notice, for instance), we are unable to exit it.

Opened #22005 . However, if I remember correctly, I think our team had a discussion about this, and decided that the UX flow would just also allow "accept" / "deny" buttons at the bottom of the "learn more" paragraphs / list of items. But, I may have dreamt it 🙃 Will definitely bring it up during our next sync.

> Pass the consent string to IAB vendors. Currently, no matter what choices the user makes, IAB vendors do not get the consent string and are unable to respect the user's choices.

Opened #22006

> Pass the CMP configuration to checkConsentHref endpoint

Opened #22007

Also, just a heads up, a lot of us are just now returning from AMP Conf. So we will probably get started on these once we get everyone back. Thanks! 😄

**Comment by jasti** [comment]
Regarding  #22006, the consumption of the consent string is made available to the ad networks on the page and it is it up to the ad network to consume the string and process it. For Google ad networks, I believe there seems to be continued work with the IAB to launch a new version of the consent string, if I'm not mistaken. 

@jawadst could you clarify the purpose for the feature: #22007?

**Comment by MaximePlancke** [comment]
Hi @zhouyx @torch2424 

I have a quick question. We need the URL of the website where we are integrated and it has to be dynamic (We can't just pass it through string in the JSON). Would it be possible to use one of your pre-built variables: https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md#amp-document-hostname inside the JSON passed through amp-consent? 

Thank you in advance

**Comment by jawadst** [comment]
@zhouyx 
Some of our publishers' pages are failing AMP validation with the message `The tag 'meta name=amp-consent-blocking' is missing or incorrect, but required by 'amp-consent [type]'.`  
There seems to be a new constraint that was recently added on the meta tag (https://github.com/ampproject/amp-wp/issues/2551). I can see a mention of it on https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/integrating-consent.md but apart from that, it does not seem to be documented anywhere (I couldn't find it on https://amp.dev/documentation/components/amp-consent/).

Would you be able to provide documentation on that meta tag? What is it used for and why is it required?

**Comment by zhouyx** [comment]
@jawadst Thank you for reporting. That's is an expected behavior. 

The `amp-consent-blocking` meta tag is required if a publisher is working with a CMP using amp-consent component.

The meta tag serves the same purpose as the `data-block-on-consent` attribute. Except that it applies the rule to all components with that tag name. 

For example 
`<meta name=amp-consent-blocking content=amp-ad>` blocks all `<amp-ad>` component on the page by default. 

Unlike publishers who create their own consent management solution, we figured it's likely that publisher won't be tagging their page with the `data-block-on-consent` attribute manually when working with the CMP. To prevent the case where there's amp-consent implemented, but no component is actually blocked by it, we decided to make this meta tag mandatory with `amp-consent [type]`. 

Good point about documentation. I'll add the doc to the publisher facing amp-consent doc as well. Thanks for bringing that up. 

Please let me know if this answers your question. Thank you.

**Comment by jawadst** [comment]
@zhouyx 
That makes sense, thanks for the explanation.
The main caveat is that vendors (and, in particular, IAB vendors), are expected to read the consent string in JavaScript so you will have websites that do not explicitly block anything with a `meta` tag or a `data-block-on-consent` attribute but still use a CMP for IAB vendors. I don't think that a page that lacks the `meta` tag is always invalid.

**Comment by zhouyx** [comment]
Thank you for clarifying. 

When you mentioned vendor, I believe you're referring to the ad vendor. As you said some ad vendors choose to only read the consent string, but not have them blocked by the consent. The behavior can be overwritten by vendors. For example, adsense has the `consentHandlingOverride: true,` in their config, so `<meta name=amp-consent-blocking content=amp-ad>` won't be able to block this vendor unless publisher specifically overwrite the behavior again. 

I understand that there are websites that do not explicitly block anything. But even that we'd like that to be explicitly declared using the meta tag `<meta name=amp-consent-blocking content>`. This is our best effort to avoid the case where publishers fail to tag their page. 

Please let us know if you think that it's a very common usage where one doesn't want to block any components on the page. Open to other design ideas. Thank you

**Comment by jawadst** [comment]
@zhouyx 
I wasn't aware that you could have a `<meta name=amp-consent-blocking content>` tag. I think that works fine in this case. More documentation on this tag that we could refer our clients to would be appreciated but the principle works well :)

Thanks for your help!

**Comment by mguerra10128** [comment]
@zhouyx Im in the process of supporting amp consent in our cmp (will be utilizing the iframe based approach), read over the following two docs:

https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/amp-consent.md
https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/integrating-consent.md

I do see in the docs any mention detailing where/how the consentString gets created and stored?

From the docs it mentions that the checkConsentHref instructs AMP to make a POST to the specified URL when the consent state is unknown. So lets say, its the first time a user is coming to the AMP page, and therefore that consent is unknown. The endpoint is hit, returns true to show the prompt. We open an iframe showing a consent message, and the user hits Accept. 

For this case, where/when does the consent string get generated and stored. If its the cmp thats creating it, can you let me know from which API that the cmp has should this be coming back in the response?

**Comment by mguerra10128** [comment]
@zhouyx we are currently blocked on the above question, any details you can share would be greatly appreciated.

**Comment by zhouyx** [comment]
Thank you @mguerra10128 for the question. 

consentString is generated by the CMP, and passed to the AMP doc from the CMP iframe using a post message. 

For example:
```
window.parent.postMessage({
  type: 'consent-response',
  action: 'accept',
  info: /string/, /* optional */
}, '*');
```

The `info` here can be used to pass the consentString. The AMP doc will handle the string and store it using the localStorage API afterwards. 

The post message API is documented at https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/integrating-consent.md#accept

Please let me know if this answers your question. Thank you!

**Comment by andresilveirah** [comment]
@zhouyx thank you for your quick reply, it was definitely helpful. I'm working with @mguerra10128 to integrate our CMP solution to AMP and there's still one piece of the consent puzzle missing. 
We're trying to pass some info to the iframe (according to https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/integrating-consent.md#client-information-passed-to-iframe) but the value of `window.name` inside the iframe is `"amp_iframe0"`. 

Here's a snippet of our implementation
```html
<amp-consent id="consent" layout="nodisplay">
  <script type="application/json">{
    "consents": {
        "publishers-site-id": {
            "checkConsentHref": "/consents/check",
            "promptUI": "consentUI"
        }
    },
    "postPromptUI": "privacy-settings-prompt"
  }</script>
  <div id="consentUI" class="popupOverlay">
      <div class="consentPopup">
          <amp-iframe
              layout="fill"
              sandbox="allow-scripts allow-same-origin allow-popups allow-popups-to-escape-sandbox"
              src='//localhost:8080/amp-ui.html'>
            <div placeholder>Loading...</div>
          </amp-iframe>
      </div>
  </div>
  <div id="privacy-settings-prompt">
    <button on="tap:consent.prompt()">Privacy Settings</button>
  </div>
</amp-consent>
```

Not sure if this is worth noticing, but we're using localhost to serve the consent ui. Are we missing something?

**Comment by zhouyx** [comment]
Hello @andresilveirah and @mguerra10128

You mentioned that you are integrating with AMP as a CMP. I have a question on how you want to work with publishers. The example you provide (with inline `<amp-iframe>`) is the solution we provide to publishers who want to manage their own consent. There are some restrictions with this approach, including not allowing consent string. 

If you are providing solution as a CMP to publishers, please try the `promptUISrc` documented [here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/integrating-consent.md#adding-your-configuration-to-amp). Another example you can refer to is the [local example ](https://github.com/ampproject/amphtml/blob/master/test/manual/amp-consent-cmp.html#L261) we built. Please let me know if that works for you. 

Re: the iframe window name. Yes that how the `<amp-iframe>` window is calculated. Which shouldn't be an issue if you are using the `promptUISrc`.
https://github.com/ampproject/amphtml/blob/7e27f4ef9eac01f48751971323325a6645a79602/extensions/amp-iframe/0.1/amp-iframe.js#L404

Thank you

**Comment by andresilveirah** [comment]
Hi @zhouyx our intention was to make sure our consent UI (which will be wrapped in the iframe) is complete and working well with the AMP APIs **before** opening a PR to change the https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/0.1/cmps.js adding our configuration. 
But it seems to me that we'll first have to add our CMP as a valid type and then make sure our code is working well, since the `<amp-iframe>` is created differently when using `<amp-consent>` with the `type` attribute, right?

**Comment by zhouyx** [comment]
Hello @andresilveirah I see, that's commonly asked question. I should definitely document it better. All configuration can be inlined. When you work on the development, you can simply inline the configuration to test locally. Could you please try inline your configuration? For example: 

```
<amp-consent id='fake' layout='nodisplay'>
      <script type="application/json">{
        "consentInstanceId": "publishers-site-id",
        "checkConsentHref": "http://localhost:8000/get-consent-v1",
        "promptUISrc": //example ui src,
        ....
      }
</script></amp-consent>
```

**Comment by andresilveirah** [comment]
Hello @zhouyx that's great! By inlining the CMP's config like you suggested I'm able to test our solution 🎉 
At the moment everything seem to be working fine with one exception. Entering fullscreen. 
In our consent UI I'm calling:
```javascript
window.parent.postMessage({ type: 'consent-ui',  action: 'ready' }, '*');
window.parent.postMessage({ type: 'consent-ui', action: 'enter-fullscreen' }, '*');
```
But the iframe is still displayed at the bottom of the screen:
<img width="720" alt="Screenshot 2019-08-08 at 12 44 27" src="https://user-images.githubusercontent.com/228650/62697179-6e891f00-b9da-11e9-8a53-88663eb69f02.png">

Not sure if this is something related to AMP or styling. And if it's styling, not sure if this is the consent ui's fault or the amp page (loading the `<amp-consent>`).

Do you have an idea? Just in case it's useful, [here's my example project](https://github.com/andresilveirah/amp-consent-response).

**Comment by MaximePlancke** [comment]
Hi @zhouyx 

We have some feedbacks about a problem with the AMP validator using amp-consent with the `type` attribute. 
Here is the issue: `The tag 'meta name=amp-consent-blocking' is missing or incorrect, but required by 'amp-consent [type]'.`
We are trying to find where this error message comes from but we can't find any documentation about it. Adding `<meta name="amp-consent-blocking" content="">` solves the problem but is there a reason for that tag to be there? 

Thank you!

**Comment by zhouyx** [comment]
Hi @andresilveirah Great to hear that the inlining configuration works. 

Just tested your example page. And I can confirm that the `enter-fullscreen` doesn't work well. 

Implementation wise, there's a delay between AMP receives the `ready` message and actually set the UI visibility to true. If the `enter-fullscreen` is received between this period, the message will be discarded. Which is what happens with your case.
However the behavior is sort of by design. The consent UI shows up without any user interaction, because of that we want to keep its initial size to be small enough to not block users. `enter-fullscreen` is an API we provide to expand the consent UI upon user click, not an API to be called along with `ready` by design. 

Please let us know your opinion on introducing an initial non blocking prompt window before entering the full screen. Thanks

**Comment by andresilveirah** [comment]
Hi @zhouyx thank you for all your support. The `enter-fullscreen` wasn't clear to me. I understand the intention behind showing the message on the bottom by default, I believe it's less intrusive than the fullscreen. However, the way we implemented our UI, the publisher will be able to choose which experience best suit their needs, hopefully the user needs will be their 1st priority :)

Regarding the blocking behaviour. If I understood correctly, just by having a `<amp-consent>` present in the page won't, by default, block ads from loading, right?
Also, once the user has accepted the consent message (thus triggering the `accept` action) it'll be up to the add vendors to consume the consent string and decide to respect that or not, is that correct?

Perhaps, if I can stretch this comment just a bit further, do you mind having a look at #23917 ?
Once again, thanks for the fast replies.

**Comment by zhouyx** [comment]
Hello @andresilveirah 

>However, the way we implemented our UI, the publisher will be able to choose which experience best suit their needs, hopefully the user needs will be their 1st priority :)

Thanks for the feedback. The decision was made along with our UX engineer after studying a list of existing CMP solutions. We are hoping to not break user experience too much by limiting the size of the consent UI. Please let us know if you can workaround with this. We're also open to any change as long as we can ensure good user experience. Thanks

>Regarding the blocking behaviour. If I understood correctly, just by having a <amp-consent> present in the page won't, by default, block ads from loading, right?

That's right. You will either need the `data-block-on-consent` attribute to block individual element on the page. Or `<meta amp-consent-blocking>` to block certain component. 

>Also, once the user has accepted the consent message (thus triggering the accept action) it'll be up to the add vendors to consume the consent string and decide to respect that or not, is that correct?

Yes and no. If an ad vendor has opt-in to handle consent, then yes. It will be up to them to consume the consent state and string. However if an ad vendor hasn't declared to handle consent, then the AMP runtime will instead handle the consent state and block the ad from rendering accordingly.

**Comment by andresilveirah** [comment]
Hi @zhouyx I have the feeling every time you answer me I come up with 2 new questions 😅 

> If an ad vendor has opt-in to handle consent, then yes. It will be up to them to consume the consent state and string. However if an ad vendor hasn't declared to handle consent, then the AMP runtime will instead handle the consent state and block the ad from rendering accordingly.
That makes sense. But what happens if the user has given consent to only _some_ vendors in the page? In our case, that'd trigger an `accept` action and the consent string would be encoded to include which vendors and purposes were consented by the user. However, if the ad vendor did not declare to handle consent, it'd be allowed to render (since the consent state is `accepted`) even if the consent string says otherwise. Right?

What about non-IAB vendors? Is there a way we could pass the consent signal to vendors which aren't included in the consent string, individually? Or do we have to stick with a binary accept all or none?

**Comment by zhouyx** [comment]
Hello @andresilveirah Those are great questions! 


> if the ad vendor did not declare to handle consent, it'd be allowed to render (since the consent state is accepted) even if the consent string says otherwise. Right?

That's right.
I tend to think the action `accept/reject` to be the generic signal to block components which don't support customized consent handling. And then the consent string to be the extra information for opt-in vendors to consume. So in the case you described I think it's best to go with  `reject` I think. 

As you also mentioned. There's a caveat here, Two vendors that don't handle consent doesn't work well together. The reason behind is that AMP doesn't want to set up its own consent string format to support such use case. 

Let me know what do you think, or if there are any good ideas on collecting consent for individual vendors.  Thanks!

**Comment by tla-sirdata** [comment]
Hi @zhouyx 

We have a CMP and would like to integrate AMP. Here is the related I2I issue : https://github.com/ampproject/amphtml/issues/24879

We started the development and everything works well so far but after having been through all the discussions and documentation, I still have some questions.

- As the consent prompt is not shown again when there is an existant consent state, how long the consent information is stored ?

- In the `window.name` passed to the iframe, I can see two attributes `consentState` and `consentStateValue` with the same value (`'accepted'/'rejected'/'unknown'`). What's the difference between them ?
And what is the `isDirty` attribute for ?

- In the `checkConsentHref` response I tried to add some key-value pairs with the `sharedData` field but I can't find how the client can read them.

- Is it possible to add the `promptUI` button outside the `amp-consent` component ? Typically, a publisher could add the button in the footer.

- In addition to IAB vendors, our CMP allows to collect consent for Google and publisher's purposes. How can we provide these information ?

- I understand that you want the initial size to be small enough to not block users. However, we have to display a very significant amount of information on the initial layer so it's better for us to use the fullscreen mode in order to keep the layer readable. Moreover, our CMP do not block users as there is always a way to close it without giving consent. What do you think about this approach?

Thanks!

**Comment by tla-sirdata** [comment]
Hi @zhouyx can you help us in integrating our CMP in AMP ? Thanks!

**Comment by zhouyx** [comment]
@tla-sirdata replied to your questions in #24879 as this thread is getting too long. Thanks

**Comment by Vasile-Peste** [comment]
@zhouyx I'm implementhing a Cookie/CMP solution, I need some clarifications:
1. In this issue you mentioned a `'forcePrompt': boolean, //optional` field that can be returned to the checkConsentHref POST request. I tried to use it because we need to load the promptUI each pageview (which is an iframe in our case: we need to load the iframe for each pageview).

2. The documentation says that the consentString is passed in the name attribute of the iframe... how does this makes sense if the iframe is loaded only if there is no consent given? If I give consent then the iframe is not loaded on the next pageviews... so you are not passing the consentString to it. When the iframe (promptUI) loads for the first time or after expireCache is set, its name attribute is just name="amp_iframe0".

Summary:
1. Is there a way to load the iframe (promptUI) each pageview, even after consent is given?
2. Is there a way to get the last known consentString from within the iframe?

Regards,
Vasile.

**Comment by zhouyx** [comment]
Hello @Vasile-Peste, happy to clarify. We've changed the API and naming a bit from the proposal here. Please refer to https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/amp-consent.md for latest doc.

>Is there a way to load the iframe (promptUI) each pageview, even after consent is given?
Yes an no. `expireCache` (which equals to the `forcePrompt` field you mentioned) would clear the stored consent, so that iframe will be load again the next user visit. However there's no way to force reload the iframe within the current visit. 

>Is there a way to get the last known consentString from within the iframe? As I explained I can't read the name attribute. I'm missing something?
Stored consentString will be passed to the iframe via name attribute, it will also be passed to `checkConsentHref` within the request body. https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/amp-consent.md#request

>Since we can return a JSON sharedData from the checkConsentHref endpoint, is there a way to access this sharedData also from withing the promptUI iframe?
Good point! There's currently no way to pass `sharedData` to the iframe. Can I ask what kind of information you want to pass from the endpoint to the iframe? `sharedData` is introduced to pass addtional information to vendors, I'm thinking maybe a good feature to have is to allow `checkConsentHref` endpoint to expand and override the `clientConfig` instead?

**Comment by zhouyx** [comment]
Hello folks, 
Wonder if the new ITP's 7-Day Cap on All Script-Writeable Storage is going to affect your product. https://webkit.org/blog/10218/full-third-party-cookie-blocking-and-more/
Thanks.

**Comment by jawadst** [comment]
@zhouyx 
There will definitely be an impact on most CMPs as we use client-side storage for keeping a local copy of the consent (just like AMP). We have a few workarounds possible: using server-side cookies (as cookies set by HTTP responses are unaffected at the moment) or login-based consent when possible being the main two options.

**Comment by zhouyx** [comment]
Thanks @jawadst 
Could you expand a bit more on the server-side cookies workaround. Will publishers need to host the server to echo the consent information via http cookies, otherwise my understanding is that there will also be the third party cookie restriction.

**Comment by saraneus** [comment]
We are creating a CMP for our publishers and would like to do an integration test locally before we actually create a pull request. How can we create a full working demo?

Our demo looks like the following so far:

`<amp-consent id="myUserConsent" layout="nodisplay">
    <script type="application/json">{
        "consentInstanceId": "consent-id",
        "consentRequired": true,
        "promptUIaaa": "consentDialog",
        "postPromptUI": "post-consent-ui",
        "checkConsentHref": "https://open-cmp-local-api.de:5003/consent/check",
        "onUpdateHref": "https://open-cmp-local-api.de:5003/consent/update",
        "promptUISrc": "https://open-cmp-local-iframe.de:5003/demo/amp/cmp-frame.html",
        "storageKey": "cmp_foo_bar"
    }</script>
    <div id="post-consent-ui">
        <button on="tap:myUserConsent.prompt()">Update Consent</button>
    </div>
</amp-consent>`

In Dev tools I can see a iframe being created loading promptUISrc, which is the standalone CMP. No errors reported. **But the iframe is hidden and only a spinner is displayed**. It's probably something I missed, could someone give me a hint?

**Comment by zhouyx** [comment]
Hello all! We're going to hold a breakout session about consent support in AMP during the online OpenJS World & OpenJs Collaborator Summit next week. (The AMP Contributor Summit has been merged into this summit this year)

We'll be covering updates on `<amp-consent>`, and are open to feedback and questions. Please register for the summit if you're interested! [Registration for the online OpenJS World & OpenJS Collab Summit](https://events.linuxfoundation.org/openjs-world/) is now open!  

Monday June 22 - OpenJS Collab Summit New Contributor Day
Tuesday June 23 and Wednesday June 24 - OpenJS World, bringing together the entire OpenJSF/JS/web community
Thursday June 25 and Friday June 26 - OpenJS Collab Summit Project Summit Days

Most of the AMP-community-specific talks and breakout sessions will be on Thursday & Friday, but everyone is welcomed/encouraged to attend all days to engage with the wider web/JS community.
When you register, make sure to indicate that you are part of the AMP community so we can get a sense of how many AMP people will be there.

The AMP talk & breakout schedule will be coming soon. :)

Thanks!

**Comment by zhouyx** [comment]
Here's the AMP talk & breakout schedule. https://docs.google.com/spreadsheets/d/1GY9qNrjxM6Ro-jFSGscL11sKQA46msRMdSrQMhTL8K0/edit#gid=461228868

**Comment by opencmp** [comment]
I'm wondering if there is a plan to support Google's Additional Consent Mode in AMP: https://support.google.com/admanager/answer/9681920
Thanks!

**Comment by zhouyx** [comment]
Thanks for the question. Yes the AMP team is aware of the additional consent string and has added support to allow CMPs to pass that to AMP via [metadata additionalConsent field](https://github.com/ampproject/amphtml/blob/master/extensions/amp-consent/integrating-consent.md#consentmetadata). 
Vendors could access the information via provided API, but we are not aware of any vendors who are utilizing the API today.
